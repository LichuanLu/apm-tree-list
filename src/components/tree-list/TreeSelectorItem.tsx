/**
 * 支持单选或者多选
 */

import React from 'react'
import ClassNames from 'classnames'
import { EmitterSubscription } from 'fbemitter'

import { List, Flex, Button, Checkbox } from 'antd-mobile'

import { TreeSelectorItemProps as TreeSelectorItemPropsType } from './PropsType'

const Item = List.Item

const isDataInSelItem = (data, selItem) => {
  if (!(data && selItem && selItem.length > 0)) {
    return false
  }
  for (let i = 0; i < selItem.length; i += 1) {
    if (selItem[i].id === data.id) {
      return true
    }
  }
  return false
}

const selected = (showSelect, selItem, rowData) => (showSelect && selItem && isDataInSelItem(rowData, selItem))

export default class TreeSelectorItem extends React.Component<TreeSelectorItemPropsType, any> {
  static defaultProps: Partial<TreeSelectorItemPropsType> = {
    prefixCls: 'amp-tree-selector-item',
  }

  subscription: EmitterSubscription
  constructor(props) {
    super(props)
    this.state = {
      selItem: [],
    }
    this.subscription = props.emitter.addListener('selEvent', (selItem) => {
      if (selItem) {
        this.setState({
          selItem,
        })
      }
    })
  }

  componentDidMount() {
    // console.log('TreeSelectorItem componentDidMount')
  }

  isDataInSelItem(data, selItem) {
    if (!(data && selItem && selItem.length > 0)) {
      return false
    }
    for (let i = 0; i < selItem.length; i += 1) {
      if (selItem[i].id === data.id) {
        return true
      }
    }
    return false
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { rowData } = this.props
    const { selItem } = this.state
    if (nextProps.rowData !== rowData) {
      return true
    }

    if (!nextState.selItem) {
      return false
    }
    // 优化性能，只刷新变化的
    if (selItem && isDataInSelItem(rowData, selItem) && !(isDataInSelItem(rowData, nextState.selItem))) {
      return true
    }

    if (selItem && !(isDataInSelItem(rowData, selItem)) && isDataInSelItem(rowData, nextState.selItem)) {
      return true
    }

    // if(this.state.selItem && this.state.selItem.id === this.props.rowID
    //   &&  nextState.selItem.id !== this.props.rowID){
    //   return true
    // }
    //
    // if((this.state.selItem || (this.state.selItem &&
    //   this.state.selItem.id !== this.props.rowID))
    //   &&  nextState.selItem.id === this.props.rowID){
    //   return true
    // }

    return false
  }

  componentWillUnmount() {
    this.subscription.remove()
  }

  render() {
    const { rowData, showSelect, onChange, toChildLevel, style, prefixCls, className } = this.props
    const sel = this.state.selItem
    const wrapCls = ClassNames(prefixCls, className)
    return (
      <Item
        className={wrapCls}
        style={style}
      >
        <Flex justify="between">
          {
            showSelect ? <Checkbox
              className={`${prefixCls}-checkbox`}
              checked={selected(showSelect, sel, rowData)}
              onChange={onChange}
            /> : ''
          }
          <span onClick={rowData.children ? toChildLevel : () => {}}>{rowData.label}</span>
          {
            rowData.children ?
              <Button className="next-level-btn" type="ghost" inline icon="right" onClick={toChildLevel} />
              :
              <Button
                className="next-level-btn"
                type="ghost"
                disabled
                inline
                icon="right"
                style={{ color: '#272727' }}
              />
          }

        </Flex>
      </Item>
    )
  }
}
