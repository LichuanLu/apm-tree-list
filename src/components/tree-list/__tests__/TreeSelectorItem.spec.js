import React from 'react';
import {render /* , mount */, shallow} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import sinon from 'sinon-sandbox';
import {EventEmitter} from 'fbemitter';
import {Checkbox} from 'antd-mobile';

import TreeSelectorItem from '../TreeSelectorItem';

describe('TreeSelectorItem', () => {
  describe('render', () => {
    const rowData = {
      label: 'TreeSelectorItem text',
      children: true,
    };

    const emitter = new EventEmitter();

    const treeSelectorItem = (
      <TreeSelectorItem rowData={rowData} emitter={emitter} />
    );

    it('render mock data correctly', () => {
      const wrapper = render(React.cloneElement(treeSelectorItem));
      expect(renderToJson(wrapper)).toMatchSnapshot();
    });

    it('contains radio button if showSelect set true', () => {
      const wrapper = shallow(
        React.cloneElement(treeSelectorItem, {showSelect: true}),
      ).dive();
      expect(wrapper.find(Checkbox).length).toBe(1);
    });

    it('childlevel button show and can be click when have children params', () => {
      const toChildLevelStub = sinon.stub();
      const wrapper = shallow(
        React.cloneElement(treeSelectorItem, {
          showSelect: true,
          toChildLevel: toChildLevelStub,
        }),
      ).dive();
      const nextLevelBtn = wrapper.find('.next-level-btn');
      expect(nextLevelBtn.length).toBe(1);
      nextLevelBtn.simulate('click');
      expect(toChildLevelStub).toHaveProperty('callCount', 1);
    });
  });
});
