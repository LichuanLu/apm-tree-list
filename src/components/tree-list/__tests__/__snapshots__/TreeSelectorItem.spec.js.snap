// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelectorItem render render mock data correctly 1`] = `
<div
  class="am-list-item amp-tree-selector-item am-list-item-middle"
>
  <div
    class="am-list-line"
  >
    <div
      class="am-list-content"
    >
      <div
        class="am-flexbox am-flexbox-justify-between am-flexbox-align-center"
      >
        <span>
          TreeSelectorItem text
        </span>
        <a
          aria-disabled="false"
          class="am-button next-level-btn am-button-ghost am-button-inline am-button-icon"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="am-button-icon am-icon am-icon-right am-icon-md"
          >
            <use
              href="#right"
            >
              
            </use>
          </svg>
        </a>
      </div>
    </div>
  </div>
  <div
    class="am-list-ripple"
    style="display:none"
  >
    
  </div>
</div>
`;
